Explorer/Assets/Scripts/ECS/SceneLifeCycle/Realm/ChangeRoomsRealmNavigator.cs

```diff
@@ -0,0 +1,33 @@
+using CommunicationData.URLHelpers;
+using Cysharp.Threading.Tasks;
+using System;
+using System.Threading;
+using UnityEngine;
+
+namespace ECS.SceneLifeCycle.Realm
+{
+    public class ChangeRoomsRealmNavigator : IRealmNavigator
+    {
+        private readonly IRealmNavigator origin;
+        private readonly Action reconnect;
+
+        public ChangeRoomsRealmNavigator(IRealmNavigator origin, Action reconnect)
+        {
+            this.origin = origin;
+            this.reconnect = reconnect;
+        }
```

Comment:
I see this approach often in your code style. I am not really sure that it worth it and that we have potential needs for it. I know, it is for better modularity. However, with it you create additional abstraction layer (by this nesting) and increases maintainability difficulty. I would rather go with the KISS approach and add it directly to the only one class that implments `IRealmNavigator` in our code base - `RealmNavigator`

Reply:
I mainly disagree that "in its core just another way of inheritance", completely opposite: It's Decorator pattern with advantages:

1. SingleResponsibility - each decorator bring its own functionality that keeps classes simple and tiny
2. OpenClosed - allows extend a behaviour without modifying already implemented behaviour, that leads to much simple refactoring
3. It's mainly the composition: it's being built once and stays immutable
4. Loose coupling - this pattern allows easily remove any piece of functionality without changing other classes, changes lives in a smaller scope. Inheritance oppositely brings high coupling: derived from some class you can't disband from it without side effecting classes that derived from the current
5. Removes a flaw: If there exists a single RealmNavigator when functionality grows the classes grows with it accordingly and becomes huge over time thats called "GodObject"
6. It allows reuse the same concept without "Configurating" a huge object. It's much harder to reuse it with a huge. As you can see I reuse same `ConnectiveRoom` in different scenarios

This pattern very similar with Builder pattern, you can find examples `WithLog` in some places in the codebase

Reply:
I see the points of both of you.

- I definitely see advantages of such "composition"
- I see it's being the common pattern you apply everywhere @NickKhalow 
- I'd propose to be a bit more selective and evaluate every scenario individually because:
    - with several levels of nesting the code becomes harder to read and perceive its purpose
    - amount of boilerplate (especially proxy-like scenarios) 
    - Be aware of `YAGNI`: don't introduce something that you don't need at the moment unless its evolution is clear
        - the most of occasions are likely `1 -> 1 -> ... 1` chain. Re-usage of intermediate classes doesn't have value or is not clear if it can be ever feasible/needed in the future

Reply:
I'm open to discuss other ways that allow to keep principles that I described above 

_with several levels of nesting the code becomes harder to read and perceive its purpose_
Yes, here we can apply the same approach that used in the DynamicWorldContainer, just separate objects by their interfaces and make them more 'Solid'

Could you bring examples where you see a violation of YAGNI? Also for amount of boilerplate pls

Reply:
My bad, sorry @NickKhalow, it is indeed mainly a Composition ("has a"), while also a bit of inheritance ("is a"). But, my point was more on YAGNI and KISS. I agree with @mikhail-dcl that we should limit use of decorator only when the dynamic flexibility they provide is necessary. And we should use simpler solutions where possible, such as simple methods or extending classes directly if the decoration is always going to be applied, as in our case (KISS). 

The main problems I see: 
- debugging and understanding difficulty (harder to track all dependencies down, similar to how it goes with deep inheritance chain)
- complex object construction
- and tiny performance concerns (since it introduces an extra layer of method calls, which can impact performance, particularly with a large number of decorators or frequent method invocations - not applied in this case of course). 

I see it in a same way as presenting interfaces everywhere - it is just a tool, not rule. We should use it wisely.

The only use-case of using provided flexibility I see is when we decide to dynamically drop/disable `MultiplayerPlugin`, but then we need also dynamic way to inject or re-inject proper `IRealmNavigator`. As for now, I don't see such use-case coming in near future and don't see any issues of presenting Decorator in future, when it will be really needed (YAGNI).

Reply:
One example that came to me is a junction of `IRoom`/`IConnectiveRoom`:

- if you don't have the context it's quite hard to understand how methods are called
   - every implementation relies on the interface implemented by itself
   - if you try to follow the calls with `Rider` they are circular because of that
- When I see such distribution of interfaces my thoughts are:
   - `IRoom` is independent from `InteriorRoom` and can be used straight-away because they implement the same interface and we keep in mind (expect applicability) **Liskov substitution principle**
      - Now I am trying to follow the calls flaw and I don't think that the app will function properly if some (all?) methods are called on the `Room` instead of `InteriorRoom` (proxy). But from the code perspective it's impossible to tell as they are hidden under the same interfaces
         - E.g. try to follow `ConnectAsync`: due to extensive usage of interfaces it's impossible to say at first glance on which type of object it will be called
      - A related question: do we need such amount of interfaces? Yes, they are properly segregated but on practice they are not used in separation, neither tested.
         - E.g. there is no usage for `ITracksHub, IRoomConnectionInfo`. And they are not testable because nothing depends on them